src/aklite_client_ext.cc

```diff
@@ -0,0 +1,143 @@
+#include "primary/reportqueue.h"
+
+#include "aktualizr-lite/tuf/tuf.h"
+#include "composeapp/appengine.h"
+#include "composeappmanager.h"
+#include "docker/restorableappengine.h"
+#include "ostree/repo.h"
+#include "tuf/akhttpsreposource.h"
+#include "tuf/akrepo.h"
+#include "tuf/localreposource.h"
+#include "uptane/exceptions.h"
+
+static Uptane::Target toUptaneTarget(const TufTarget& t) {
+  Json::Value target_json;
+  target_json["hashes"]["sha256"] = t.Sha256Hash();
+  target_json["custom"]["targetFormat"] = "OSTREE";
+  target_json["length"] = 0;
+  Uptane::Target target(t.Name(), target_json);
+  return target;
+}
+
+// Returns the target that should be installed, if any.
+// It might be an updated version, a rollback target, or even the currently installed target, in case we need to sync
+// apps
+GetTargetToInstallResult AkliteClientExt::GetTargetToInstall(const LocalUpdateSource* local_update_source, int version,
+                                                             const std::string& target_name, bool allow_bad_target,
+                                                             bool force_apps_sync) {
+  // Tell CheckIn and CheckInLocal methods to not call notifyTufUpdateFinished in case of success
+  usingUpdateClientApi = true;
+  client_->setAppsNotChecked();
+  bool rollback_operation = false;
+
+  CheckInResult checkin_res = local_update_source == nullptr ? CheckIn() : CheckInLocal(local_update_source);
+
+  if (!checkin_res) {
+    LOG_WARNING << "Unable to update latest metadata";
+    return GetTargetToInstallResult(checkin_res);
+  }
+
+  auto candidate_target = checkin_res.SelectTarget(version, target_name);
+  if (candidate_target.IsUnknown()) {
+    std::string err = "No matching target";
+    client_->notifyTufUpdateFinished(err);
+    LOG_WARNING << err;
+    return {GetTargetToInstallResult::Status::TufTargetNotFound, TufTarget(), err};
+  }
+
+  const auto current = GetCurrent();
+  if (IsRollback(current) && current.Name() == candidate_target.Name()) {
+    // Handle the case when Apps failed to start on boot just after an update.
+    // This is only possible with `pacman.create_containers_before_reboot = 0`.
+    LOG_INFO << "The currently booted Target is a failing Target, finding Target to rollback to...";
+    const TufTarget rollback_target = Target::toTufTarget(client_->getRollbackTarget());
+    if (rollback_target.IsUnknown()) {
+      const auto err{boost::str(boost::format("Failed to find Target to rollback to after a failure to start "
+                                              "Apps at boot on a new version of sysroot;"
+                                              " failing current Target: %s, hash: %s") %
+                                current.Name() % current.Sha256Hash())};
+
+      LOG_ERROR << err;
+      client_->notifyTufUpdateFinished(err);
+      // TODO: Specify proper error status
+      return {GetTargetToInstallResult::Status::Failed, TufTarget(), err};
```

Comment:
RollbackTargetNotFound?

Reply:
Added `RollbackTargetNotFound` return code.

Reply:
Not sure I see it.

Reply:
Is at this same location.

https://github.com/foundriesio/aktualizr-lite/blob/f2e2221f071125e2f32de4e1784aca86f863a36a/src/aklite_client_ext.cc#L66